src/frame/decompress.rs

```diff
@@ -1 +1,312 @@
+use std::{fmt, hash::Hasher, io, mem::size_of};
+use twox_hash::XxHash32;
 
+use super::header::{BlockInfo, BlockMode, FrameInfo, MAX_FRAME_INFO_SIZE, MIN_FRAME_INFO_SIZE};
+use super::Error;
+use crate::block::WINDOW_SIZE;
+
+/// A reader for decompressing the LZ4 frame format
+///
+/// This Decoder wraps any other reader that implements `io::Read`.
+/// Bytes read will be decompressed according to the [LZ4 frame format](
+/// https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md).
+///
+/// This reader can potentially make many small reads from the underlying
+/// stream depending on its format, therefore, passing in a buffered reader
+/// may be beneficial.
+pub struct FrameDecoder<R: io::Read> {
+    /// The underlying reader.
+    r: R,
+    /// Whether we've read the a stream header or not.
+    /// Also cleared once frame end marker is read and Ok(0) is returned.
+    frame_info: Option<FrameInfo>,
+    /// Xxhash32 used when content checksum is enabled.
+    content_hasher: XxHash32,
+    /// Total length of decompressed output for the current frame.
+    content_len: u64,
+    /// The compressed bytes buffer, taken from the underlying reader.
+    src: Vec<u8>,
+    /// The decompressed bytes buffer. Bytes are decompressed from src to dst
+    /// before being passed back to the caller.
+    dst: Vec<u8>,
+    /// Index into dst and length: starting point of bytes previously output
+    /// that are still part of the decompressor window.
+    ext_dict_offset: usize,
+    ext_dict_len: usize,
+    /// Index into dst: starting point of bytes not yet given back to caller.
+    dst_start: usize,
+    /// Index into dst: ending point of bytes not yet given back to caller.
+    dst_end: usize,
+}
+
+impl<R: io::Read> FrameDecoder<R> {
+    /// Creates a new Decoder for the specified reader.
+    pub fn new(rdr: R) -> FrameDecoder<R> {
+        FrameDecoder {
+            r: rdr,
+            src: Default::default(),
+            dst: Default::default(),
+            ext_dict_offset: 0,
+            ext_dict_len: 0,
+            dst_start: 0,
+            dst_end: 0,
+            frame_info: None,
```

Comment:
I think it would be better read_frame_info in the constructor and make frame_info non optional

Reply:
I think it's a good practice that io::Read/Write wrappers are lazy so they don't do much unless read/write is called. 
The FrameInfo being present/not-present is important so an FrameDecoder can read multiple frames from the underlying io::Read (like in tests.rs `frame::concatenated`).

Frame info doesn't necessarily have to be an Option I guess. I could be set to the default value (and then read from the io::Read) and then we add a boolean or something to track if it's between a frame start and frame end. That's probably less intuitive though.

Reply:
I think it's okay to have FrameInfo as Option to have lazy readers, but the information that we are at the end is stored implicitly in the frame info option, I think it's always better to be explicit.

Reply:
Maybe renaming the field to `current_frame_info` helps?

Reply:
yes I think that's better